File: javascript/packages/highlighter/src/inline-diagnostic-renderer.ts
import { colorize, hyperlink, severityColor } from "./color.js"
import { TextFormatter } from "./text-formatter.js"
import { LineWrapper } from "./line-wrapper.js"
import * as gutter from "./gutter.js"
import { DIAGNOSTIC_SEVERITIES } from "@herb-tools/core"
import { computeDiagnosticMarkers } from "./diagnostic-markers.js"

import type { Diagnostic, DiagnosticSeverity } from "@herb-tools/core"
import type { SyntaxRenderer } from "./syntax-renderer.js"
import type { DiagnosticMarker } from "./diagnostic-markers.js"

interface LineMarker {
  diagnostic: Diagnostic
  marker: DiagnosticMarker
  isLastLine: boolean
}

export class InlineDiagnosticRenderer {
  private syntaxRenderer: SyntaxRenderer

  constructor(syntaxRenderer: SyntaxRenderer) {
    this.syntaxRenderer = syntaxRenderer
  }

  private getSeverityText(severity: DiagnosticSeverity): string {
    return colorize(colorize(severity, severityColor(severity)), "bold")
  }

  private getHighestSeverity(diagnostics: Diagnostic[]): DiagnosticSeverity {
    for (const severity of DIAGNOSTIC_SEVERITIES) {
      if (diagnostics.some(diagnostic => diagnostic.severity === severity)) {
        return severity
      }
    }

    return "warning"
  }

  render(
    path: string,
    content: string,
    diagnostics: Diagnostic[],
    showLineNumbers = true,
    wrapLines = false,
    maxWidth = LineWrapper.getTerminalWidth(),
    truncateLines = false,
    codeUrlBuilder?: (code: string) => string,
  ): string {
    const highlightedContent = this.syntaxRenderer.highlight(content)
    const contentLines = content.split("\n")

    const markersByLine = new Map<number, LineMarker[]>()
    for (const diagnostic of diagnostics) {
      const markers = computeDiagnosticMarkers(diagnostic.location, contentLines)

      markers.forEach((marker, index) => {
        if (!markersByLine.has(marker.line)) {
          markersByLine.set(marker.line, [])
        }

        markersByLine.get(marker.line)!.push({
          diagnostic,
          marker,
          isLastLine: index === markers.length - 1,
        })
      })
    }

    const severityOrder: Record<DiagnosticSeverity, number> = {
      "error": 0,
      "warning": 1,
      "info": 2,
      "hint": 3
    }

    for (const lineMarkers of markersByLine.values()) {
      lineMarkers.sort((a, b) => {
        const orderA = severityOrder[a.diagnostic.severity] ?? 99
        const orderB = severityOrder[b.diagnostic.severity] ?? 99
        return orderA - orderB
      })
    }

    const lines = highlightedContent.split("\n")
    let output = showLineNumbers ? `${colorize(path, "cyan")}\n\n` : ""
    let previousLineHadMessages = false

    for (let i = 1; i <= lines.length; i++) {
      const line = lines[i - 1] || ""
      const lineMarkers = markersByLine.get(i) || []
      const lineDiagnostics = lineMarkers.map(({ diagnostic }) => diagnostic)
      const hasDiagnostics = lineMarkers.length > 0
      const hasMessages = lineMarkers.some(({ isLastLine }) => isLastLine)

      if (previousLineHadMessages) {
        output += showLineNumbers ? `${gutter.pointerPrefix()}\n` : "\n"
      }

      const highestSeverity = this.getHighestSeverity(lineDiagnostics)
      const lineColor = severityColor(highestSeverity)

      const displayLine = line
      let contentWidth = maxWidth

      if (wrapLines && showLineNumbers) {
        const prefix = gutter.linePrefix(i, hasDiagnostics, hasDiagnostics ? lineColor : undefined)
        contentWidth = gutter.availableWidth(maxWidth)

        const wrappedLines = LineWrapper.wrapLine(displayLine, contentWidth, "")

        for (let j = 0; j < wrappedLines.length; j++) {
          if (j === 0) {
            output += `${prefix}${wrappedLines[j]}\n`
          } else {
            output += `${gutter.continuationPrefix()}${wrappedLines[j]}\n`
          }
        }
      } else if (truncateLines && showLineNumbers) {
        const prefix = gutter.linePrefix(i, hasDiagnostics, hasDiagnostics ? lineColor : undefined)
        contentWidth = gutter.availableWidth(maxWidth)

        const truncatedLine = LineWrapper.truncateLine(displayLine, contentWidth)
        output += `${prefix}${truncatedLine}\n`
      } else if (showLineNumbers) {
        output += `${gutter.linePrefix(i, hasDiagnostics, hasDiagnostics ? lineColor : undefined)}${displayLine}\n`
      } else if (wrapLines) {
        contentWidth = maxWidth
        const wrappedLines = LineWrapper.wrapLine(displayLine, maxWidth)
        for (const wrappedLine of wrappedLines) {
          output += `${wrappedLine}\n`
        }
      } else if (truncateLines) {
        const truncatedLine = LineWrapper.truncateLine(displayLine, maxWidth)
        output += `${truncatedLine}\n`
      } else {
        output += `${displayLine}\n`
      }

      if (hasDiagnostics) {
        for (const { diagnostic, marker, isLastLine } of lineMarkers) {
          const pointerLength = Math.max(1, marker.end - marker.start)
          const pointer = colorize(
            "~".repeat(pointerLength),
            severityColor(diagnostic.severity),
          )

          const severityText = this.getSeverityText(diagnostic.severity)
          const diagnosticIdText = diagnostic.code || "-"
          const diagnosticId = codeUrlBuilder && diagnostic.code ? hyperlink(diagnosticIdText, codeUrlBuilder(diagnostic.code)) : diagnosticIdText
          const highlightedMessage = TextFormatter.highlightBackticks(diagnostic.message)
          const diagnosticText = `[${severityText}] ${highlightedMessage} (${diagnosticId})`
          const dimmedDiagnosticText =
            TextFormatter.dimAnsiCodes(diagnosticText)

          if (showLineNumbers) {
            const pointerPrefix = gutter.pointerPrefix()
            const pointerSpacing = " ".repeat(Math.max(0, marker.start + 1))

            output += `${pointerPrefix}${pointerSpacing}${pointer}\n`

            if (isLastLine) {
              output += `${pointerPrefix}${pointerSpacing}${dimmedDiagnosticText}\n`
            }
          } else {
            const pointerSpacing = " ".repeat(Math.max(0, marker.start))

            output += `${pointerSpacing}${pointer}\n`

            if (isLastLine) {
              output += `${dimmedDiagnosticText}\n`
            }
          }
        }
      }

      previousLineHadMessages = hasMessages
    }

    return output.trimEnd()
  }
}

File: rust/herb-highlighter/src/inline_diagnostic_renderer.rs
use std::fmt::Write as _;

use std::collections::HashMap;

use crate::color::{colorize, hyperlink, severity_color, NamedColor};
use crate::diagnostic::{Diagnostic, DiagnosticSeverity, DIAGNOSTIC_SEVERITIES};
use crate::diagnostic_markers::{compute_diagnostic_markers, DiagnosticMarker};
use crate::file_renderer::RenderOptions;
use crate::gutter;
use crate::line_wrapper::LineWrapper;
use crate::syntax_renderer::SyntaxRenderer;
use crate::text_formatter::TextFormatter;

struct LineMarker<'a> {
  diagnostic: &'a Diagnostic,
  marker: DiagnosticMarker,
  is_last_line: bool,
}

pub type CodeUrlBuilder<'a> = &'a dyn Fn(&str) -> String;

pub struct InlineDiagnosticRenderer<'a> {
  syntax_renderer: &'a SyntaxRenderer,
}

impl<'a> InlineDiagnosticRenderer<'a> {
  pub fn new(syntax_renderer: &'a SyntaxRenderer) -> Self {
    Self { syntax_renderer }
  }

  fn severity_text(&self, severity: DiagnosticSeverity) -> String {
    colorize(&colorize(severity.as_str(), severity_color(severity)), NamedColor::Bold)
  }

  fn highest_severity(&self, diagnostics: &[&Diagnostic]) -> DiagnosticSeverity {
    for severity in DIAGNOSTIC_SEVERITIES {
      if diagnostics.iter().any(|diagnostic| diagnostic.severity == severity) {
        return severity;
      }
    }

    DiagnosticSeverity::Warning
  }

  pub fn render(&self, path: &str, content: &str, diagnostics: &[Diagnostic], options: &RenderOptions, code_url_builder: Option<CodeUrlBuilder<'_>>) -> String {
    let highlighted_content = self.syntax_renderer.highlight(content);
    let content_lines: Vec<&str> = content.split('\n').collect();

    let mut markers_by_line: HashMap<usize, Vec<LineMarker<'_>>> = HashMap::new();

    for diagnostic in diagnostics {
      let markers = compute_diagnostic_markers(&diagnostic.location, &content_lines);
      let last_index = markers.len() - 1;

      for (index, marker) in markers.into_iter().enumerate() {
        markers_by_line.entry(marker.line).or_default().push(LineMarker {
          diagnostic,
          marker,
          is_last_line: index == last_index,
        });
      }
    }

    for line_markers in markers_by_line.values_mut() {
      line_markers.sort_by_key(|line_marker| line_marker.diagnostic.severity.order());
    }

    let lines: Vec<&str> = highlighted_content.split('\n').collect();

    let mut output = if options.show_line_numbers {
      format!("{}\n\n", colorize(path, NamedColor::Cyan))
    } else {
      String::new()
    };
    let mut previous_line_had_messages = false;

    for number in 1..=lines.len() {
      let line = lines.get(number - 1).copied().unwrap_or("");
      let empty: Vec<LineMarker<'_>> = Vec::new();
      let line_markers = markers_by_line.get(&number).unwrap_or(&empty);
      let line_diagnostics: Vec<&Diagnostic> = line_markers.iter().map(|line_marker| line_marker.diagnostic).collect();
      let has_diagnostics = !line_markers.is_empty();
      let has_messages = line_markers.iter().any(|line_marker| line_marker.is_last_line);

      if previous_line_had_messages {
        if options.show_line_numbers {
          let _ = writeln!(output, "{}", gutter::pointer_prefix());
        } else {
          output.push('\n');
        }
      }

      let highest_severity = self.highest_severity(&line_diagnostics);
      let line_color = severity_color(highest_severity);

      let display_line = line;

      if options.wrap_lines && options.show_line_numbers {
        let line_prefix = gutter::line_prefix(number, has_diagnostics, has_diagnostics.then_some(line_color));
        let available_width = gutter::available_width(options.max_width);

        let wrapped_lines = LineWrapper::wrap_line(display_line, available_width, "");

        for (index, wrapped_line) in wrapped_lines.iter().enumerate() {
          if index == 0 {
            let _ = writeln!(output, "{line_prefix}{wrapped_line}");
          } else {
            let _ = writeln!(output, "{}{wrapped_line}", gutter::continuation_prefix());
          }
        }
      } else if options.truncate_lines && options.show_line_numbers {
        let line_prefix = gutter::line_prefix(number, has_diagnostics, has_diagnostics.then_some(line_color));
        let available_width = gutter::available_width(options.max_width);

        let truncated_line = LineWrapper::truncate_line(display_line, available_width);

        let _ = writeln!(output, "{line_prefix}{truncated_line}");
      } else if options.show_line_numbers {
        let line_prefix = gutter::line_prefix(number, has_diagnostics, has_diagnostics.then_some(line_color));

        let _ = writeln!(output, "{line_prefix}{display_line}");
      } else if options.wrap_lines {
        for wrapped_line in LineWrapper::wrap_line(display_line, options.max_width, "") {
          let _ = writeln!(output, "{wrapped_line}");
        }
      } else if options.truncate_lines {
        let _ = writeln!(output, "{}", LineWrapper::truncate_line(display_line, options.max_width));
      } else {
        let _ = writeln!(output, "{display_line}");
      }

      if has_diagnostics {
        for LineMarker {
          diagnostic,
          marker,
          is_last_line,
        } in line_markers
        {
          let pointer_length = (marker.end - marker.start).max(1);
          let pointer = colorize(&"~".repeat(pointer_length), severity_color(diagnostic.severity));

          let severity_text = self.severity_text(diagnostic.severity);
          let diagnostic_id_text = diagnostic.code.clone().unwrap_or_else(|| "-".to_string());

          let diagnostic_id = match (code_url_builder, &diagnostic.code) {
            (Some(builder), Some(code)) => hyperlink(&diagnostic_id_text, &builder(code)),
            _ => diagnostic_id_text,
          };

          let highlighted_message = TextFormatter::highlight_backticks(&diagnostic.message);
          let diagnostic_text = format!("[{severity_text}] {highlighted_message} ({diagnostic_id})");
          let dimmed_diagnostic_text = TextFormatter::dim_ansi_codes(&diagnostic_text);

          if options.show_line_numbers {
            let pointer_prefix = gutter::pointer_prefix();
            let pointer_spacing = " ".repeat(marker.start + 1);

            let _ = writeln!(output, "{pointer_prefix}{pointer_spacing}{pointer}");

            if *is_last_line {
              let _ = writeln!(output, "{pointer_prefix}{pointer_spacing}{dimmed_diagnostic_text}");
            }
          } else {
            let pointer_spacing = " ".repeat(marker.start);

            let _ = writeln!(output, "{pointer_spacing}{pointer}");

            if *is_last_line {
              let _ = writeln!(output, "{dimmed_diagnostic_text}");
            }
          }
        }
      }

      previous_line_had_messages = has_messages;
    }

    output.trim_end().to_string()
  }
}
